javascript/ql/src/semmle/javascript/ApiGraphs.qll

```diff
@@ -21,30 +21,34 @@ module API {
    */
   class Node extends Impl::TApiNode {
     /**
-     * Gets a data-flow node corresponding to a use of the API component represented by this node.
+     * Gets a `SourceNode` corresponding to a use of the API component represented by this node.
      *
      * For example, `require('fs').readFileSync` is a use of the function `readFileSync` from the
      * `fs` module, and `require('fs').readFileSync(file)` is a use of the result of that function.
      *
      * As another example, in the assignment `exports.plusOne = (x) => x+1` the two references to
      * `x` are uses of the first parameter of `plusOne`.
+     *
+     * Note: The result from this predicate is always a `DataFlow::SourceǸode`, use `getAUse()` if
+     * you want to follow purely local data-flow and get all `DataFlow::Node`s that corrospond to a
+     * use of this API node.
      */
-    DataFlow::Node getAUse() {
-      exists(DataFlow::SourceNode src | Impl::use(this, src) |
-        Impl::trackUseNode(src).flowsTo(result)
-      )
+    DataFlow::SourceNode getAReference() {
+      exists(DataFlow::SourceNode src | Impl::use(this, src) | result = Impl::trackUseNode(src))
```

Comment:
Hm, as you say this is now becoming very similar to `getAUse`. Perhaps a bit too similar. The nice thing about the previous implementation (without `trackUseNode`) was that it had a clear correspondence to `getARhs()`, but this is now lost.

How about reverting to the previous definition without `trackUseNode`, and instead moving that ito `getACall`? The only other case where we need it seems to be a single use of `getAReference().getAnInstantiation()`; perhaps we can introduce a helper predicate analogous to `getACall` to cover that.

Reply:
Actually, looking at this again, can you remind me why we need the type tracking at all? Couldn't we remove it, and then implement `getACall` like this?

```ql
DataFlow::CallNode getACall() {
  result = getResult().getAReference()
}
```

In other words, instead of dropping out of API graphs early, then do type tracking and local flow to find a call, why not use API graphs to find the call directly, and then drop out without further flow?

Reply:
> Actually, looking at this again, can you remind me why we need the type tracking at all? Couldn't we remove it, and then implement getACall like this?
> 
> DataFlow::CallNode getACall() {
>   result = getResult().getAReference()
> }

I got the same idea last night.   
Yes, I think that would work.   

(And you meant `getReturn()` instead of `getResult()`, right?).

Reply:
> And you meant getReturn() instead of getResult(), right?

Yes, totally! (Where is my context-sensitive autocomplete in the browser?)

Reply:
I think you are better at writing those qldocs than I am.   
I've given it a go now, but you are very welcome to make corrections.